### ISSUE TITLE
Create mock bled112 process for simulating high volume broadcast traffic.

### ISSUE THREAD

**Issue by timburke** [iotile/coretools#863]
There is starter code for this [here](https://github.com/iotile/coretools/blob/master/transport_plugins/bled112/test/util/mock_bled112.py).  It is used internally for unit testing the bled112 device adapter.  

The goal of this issue should be a class that can generate high volume advertisement traffic and a command line script that takes a virtual serial port and acts like a bled112 dongle on that port.

Only a minimal amount of functionality is needed:
- set scan parameters: if active scan is true, then 2 advertisements are sent per "simulated device", one advertisement and one scan response packet, each 31 bytes.
- start scan
- stop scan
- query system.  This command is send when the bled112 device adapter turns on to figure out if any ble connections are active.

The command line script should log status periodically on what volume of broadcasts it is generating so that it's easy to see if the rate drops as we add additional things on the receiving side that could slow it down.

**Comment by DimaWittmann** [comment]
Hi Tim! 
I am trying to make a complete picture in my head.

As I understood so far, I need to create a shell with a few commands (set parameters, start, stop, query).
As a argument this script accept path to device node.
I need to implement a class that will imitate a bled112 device that receives multiple adv packets (it jsut generate adv packets) and propagates it to gateway.

The gateway is expected to start and run independently of the script, we just know it should be connected to the other side of serial port.

The first question is how scan parameters should be propagated to gateway? It looks like gateway choose between active and passive scan.
**Start scan** cmd should make the mock to generate events with adv packets, provided that connection to bled adapter is established, shouldn't they?  
**Query system**, I haven't yet understood the purpose of this cmd. Bled device adapter initialize the connection to mock, start scan and than mock can only decide whether to send scan events or not. Please share any details if you can. 

Could you verify that my vision meets your expectation? I would be happy to see any additional thoughts  from you.

**Comment by timburke** [comment]
I believe that's mostly in line with what I was thinking with the following caveats and some clarifications.

The `mock-bled112` script should be fully automatic in the sense that you start with a few command line arguments to specify how often to send advertisements (after the gateway calls `start scan`).  For example:

```
> mock-bled112 --max-advertisements-per-second 1000 --port /dev/side/b/of/virtual/tty
INFO: Starting emulated bled112 dongle serving 1000 messages per second
```

That should be the only interaction you have with the `mock-bled112` script.  It internally goes into a loop reading commands from `/dev/side/b/of/virtual/tty`, which has the other end of the serial port connected to the `iotile-gateway`.  As part of its operation, `iotile-gateway`, through is `BLED112Adapter` plugin will send `start scan` and `query system` commands over the serial port to emulated bled112 script that it needs to respond to over the same serial port.

If you look at the API reference for the (real) bled112's serial interface, you'll see a huge pile of commands.  **Very few of them are needed** for the actual use case we have for the emulator because most are related to connecting to a device and doing ble reads and writes (which the emulator will not support).

The minimum set of commands that needs to be supported are just the ones that the BLED112Adapter class sends before it gets to the part of its code where it just waits to receive advertisements.  Those are:

- `query_system`: This returns the state of the bled112 including how many active connections there are.  It should report 0 active connections.  It is called at boot time by BLED112Adapter to see if there are any connections that were left dangling from a previous session and if so they are closed.  Since none are reported, it doesn't send any `disconnect` commands.
- `set_scan_parameters`: After finishing `query_system`, BLED112Adapter then tells the bled112dongle what parameters to use for scanning.  The command must succeed but the parameters are irrelevant and can be ignored except for the `active-scan` parameter.  That must be remembered so that the appropriate length advertisements are sent.
- `start_scan`:  This is the trigger that should make the mock bled112 start aggressively sending advertisements.  If active scan was true, the advertisements are 62 bytes long.  If false, then they are 31 bytes long.  (active scan includes an additional 31 byte scan response packet).

Basically the goal is to implement just enough of the bled112 serial api to get the `BLED112Adapter` class to be able to receive advertisements.  Then we want to push advertisements including a broadcast reading as fast as possible and see how much throughput we get through the whole `iotile-gateway` system.

**Comment by timburke** [comment]
@DimaWittmann Let me know what additional questions you have based on those details.

**Comment by mattrunchey** [comment]
Closed by #869
